src/main/python/org/broadinstitute/hellbender/gcnvkernel/inference/deterministic_annealing.py

```diff
@@ -44,12 +50,70 @@ def __init__(self,
                  temperature=None):
 
         assert temperature is not None, "Temperature (a scalar theano shared tensor) is not provided"
+        approx = DeterministicMeanField(local_rv=local_rv,
+                                        model=model,
+                                        cost_part_grad_scale=cost_part_grad_scale,
+                                        scale_cost_to_minibatch=scale_cost_to_minibatch,
+                                        random_seed=random_seed,
+                                        start=start)
         super().__init__(
-            KLThermal, MeanField, None,
-            local_rv=local_rv,
-            model=model,
-            cost_part_grad_scale=cost_part_grad_scale,
-            scale_cost_to_minibatch=scale_cost_to_minibatch,
-            random_seed=random_seed,
-            start=start,
-            op_kwargs={'temperature': temperature})
+            KLThermal, approx, None, temperature=temperature)
+
+class DeterministicApproximation(Approximation):
+
+    def __init__(self, groups, model=None):
+        self._scale_cost_to_minibatch = theano.shared(np.int8(1))
+        model = modelcontext(model)
+        if not model.free_RVs:
+            raise TypeError('Model does not have FreeRVs')
+        self.groups = list()
```

Comment:
Just curious, is most of this code essentially borrowed from PyMC4 (as you referenced in the Slack thread above), or is it original?